Multiple fix improves to build local

rodrigosiqueira opened the pull request:
@melissawen pointed out that kw deploy local was not working as expected; this commit is a set of small changes that aim to address all issues associated with local deploy.

davidbtadokoro reviewed (changes requested):
Hi @rodrigosiqueira, thanks for this great change!
Everything looks legitimate for me, just left some comments.

However, when testing a local deploy in a fresh installed debian VM, I encountered one of the errors that you addressed in my first try:

<img width="818" alt="Screen Shot 2023-09-12 at 17 08 27" src="https://github.com/kworkflow/kworkflow/assets/95055784/52b58974-f03a-4469-ab89-819a793ee6c2">

But then, in the second try it worked:

<img width="1002" alt="Screen Shot 2023-09-12 at 17 08 35" src="https://github.com/kworkflow/kworkflow/assets/95055784/f4e49f7c-b443-453b-bd75-2f304cc86155">

Don't know if its a problem on my side, but appart from that, everything looks good 👍

davidbtadokoro commented:
Just an update.

I restored my VM some times and it seems that the first try always fails:

<img width="958" alt="Screen Shot 2023-09-12 at 17 17 20" src="https://github.com/kworkflow/kworkflow/assets/95055784/e48865c6-ddae-4354-887e-3d2e8400e9df">

Note: The "fail" I'm reffering is the error message and the hang that follows (I have to interrupt the execution).

melissawen commented:
> Just an update.
> 
> I restored my VM some times and it seems that the first try always fails:
> Note: The "fail" I'm reffering is the error message and the hang that follows (I have to interrupt the execution).

The same error happens here:
```
E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```

But as I already have all dependencies installed, it doesn't affect the local deployment here.

A part from that, the other issues are solved

rodrigosiqueira commented on src/plugins/kernel_install/utils.sh:
```diff
@@ -676,7 +676,7 @@ function install_kernel()
   # is why we use tee. Also note that the stdin is passed to the eval
   # inside cmd_manager.
-  cmd="grep -Fxq $name $INSTALLED_KERNELS_PATH"
+  cmd="${sudo_cmd}grep -Fxq $name $INSTALLED_KERNELS_PATH"
```
Actually, the other parts of this code need to be corrected since we should not add spaces in this string concatenation. Check `sudo_cmd='sudo -E '$` at the beginning of this function.

davidbtadokoro replied:
Oh, I see it now. It also makes sense as in case `sudo_cmd` is empty, there won't be a leading whitespace in the command.

rodrigosiqueira commented on src/plugins/kernel_install/utils.sh:
```diff
@@ -676,7 +676,7 @@ function install_kernel()
   # is why we use tee. Also note that the stdin is passed to the eval
   # inside cmd_manager.
-  cmd="grep -Fxq $name $INSTALLED_KERNELS_PATH"
+  cmd="${sudo_cmd}grep -Fxq $name $INSTALLED_KERNELS_PATH"
```
Done

rodrigosiqueira commented on tests/deploy_test.sh:
```diff
@@ -336,18 +336,22 @@ function test_update_status_log()
   local expected_data
```
done

rodrigosiqueira commented on tests/deploy_test.sh:
```diff
@@ -336,18 +336,22 @@ function test_update_status_log()
+  cmd="\"printf '%s;%s\n' '3' 'TEST_MODE' >> ${KW_STATUS_BASE_PATH}/kw_status\""
   output=$(update_status_log 3 'TEST_MODE')
 
   assert_equals_helper 'Status file remote' "$LINENO" "$ssh_prefix $cmd" "$output"
```
Done

rodrigosiqueira commented on tests/deploy_test.sh:
```diff
@@ -336,18 +336,22 @@ function test_update_status_log()
-  update_status_log 2
-  output=$(cat "$SHUNIT_TMPDIR/kw_status")
-  expected_data='2;12/31/2021-09:49:21'
+function test_update_status_log_local()
```
Done

rodrigosiqueira commented on tests/deploy_test.sh:
```diff
@@ -336,18 +336,22 @@ function test_update_status_log()
+{
+  local output
+  local expected_cmd
+  local log_date
```
Done

rodrigosiqueira commented on src/plugins/kernel_install/utils.sh:
```diff
@@ -45,10 +45,12 @@ function cmd_manager()
 
 function command_exists()
 {
-  local command="$1"
-  local package=${command%% *}
+  local cmd="$1"
+  local package=${cmd%% *}
 
   if [[ ! -x "$(command -v "$package")" ]]; then
+    # Fallback
+    [[ -x "/usr/sbin/${package}" ]] && return 0
     return 22 # EINVAL
   fi
   return 0
```
This copy-and-paste solution is a terrible solution. We need to copy it because in the remote deploy kw may not be installed in the target machine. In the future, we need to find a way to send that code without duplicating it.

rodrigosiqueira commented:
Hi @davidbtadokoro @melissawen ,
Thanks a lot for testing and reviewing this PR. I addressed most of the comments and updated this PR to solve the bugs reported by both of you. Please give it another try when you have time. If it still fails, please also add the --verbose flag and post the output here.

davidbtadokoro reviewed (approved):
HI @rodrigosiqueira, thanks so much for the update!
Everything looks good to me.

Reviewed-by: David Tadokoro \<davibtadokoro@usp.br\>

davidbtadokoro commented:
> Just an update.
> 
> I restored my VM some times and it seems that the first try always fails:
> 

Also, re-tested this behaviour and it works to me 👍

rodrigosiqueira commented:
Thanks a lot for your review and feedback.
